File: data_structures/stacks/stack.js
// Author: Austin Crane
// this is a node program, you can run this with node stack.js

// Stack simple stack implementation
// accessed via a "private" items array
function Stack() {
  this._items = [];
  return this;
}

// push adds an item onto the stack
Stack.prototype.push = function(n) {
  this._items.push(n);
}

// pop removes the item on the top of the stack
Stack.prototype.pop = function() {
  if (this.size() === 0) {
    // if your trying to access an empty stack that
    // would be considered an error
    console.error('reached end of stack');
    return;
  }
  return this._items.pop();
}

// size gives the size of the stack
Stack.prototype.size = function() {
  return this._items.length;
}

// simple "main" program to work with a stack
function stackExample() {
  var s = new Stack()
  s.push(1);
  s.push(2);
  s.push(3);

  // print size of stack
  console.log(s.size());

  // 3
  console.log(s.pop());
  // 2
  console.log(s.pop());
  // 1
  console.log(s.pop());

  // this will log an error reach end of stack
  console.log(s.pop());
}

stackExample();

File: data_structures/stacks/Python3/stack.py
#!/usr/bin/python3

class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[-1]

    def size(self):
        return len(self.items)

if __name__ == "__main__":
    stack = Stack()
    stack.push(1)
    stack.push(2)

    print("Size of stack: %d " % stack.size())
    print("Stack.peek: %d" % stack.peek())
    

    print("Stack.pop: %d" % stack.pop())
    print("Stack.pop: %d" % stack.pop())
